src/wm/title.rs

```diff
@@ -0,0 +1,126 @@
+use std::{fmt::Display, path::PathBuf};
+
+use anyhow::{anyhow, bail};
+
+use url::Url;
+
+/// Normalized wikipedia article title that can compare:
+/// - titles `Spatial Database`
+/// - urls `https://en.wikipedia.org/wiki/Spatial_database#Geodatabase`
+/// - osm-style tags `en:Spatial Database`
+///
+/// ```
+/// use om_wikiparser::wm::Title;
+///
+/// let title = Title::from_title("Article Title", "en").unwrap();
+/// let url = Title::from_url("https://en.wikipedia.org/wiki/Article_Title#Section").unwrap();
+/// let mobile = Title::from_url("https://en.m.wikipedia.org/wiki/Article_Title#Section").unwrap();
+/// let url_tag1 = Title::from_osm_tag("https://en.m.wikipedia.org/wiki/Article_Title#Section").unwrap();
+/// let url_tag2 = Title::from_osm_tag("de:https://en.m.wikipedia.org/wiki/Article_Title#Section").unwrap();
+/// assert_eq!(url, title);
+/// assert_eq!(url, mobile);
+/// assert_eq!(url, url_tag1);
+/// assert_eq!(url, url_tag2);
+///
+/// assert!(Title::from_url("https://en.wikipedia.org/not_a_wiki_page").is_err());
+/// assert!(Title::from_url("https://wikidata.org/wiki/Q12345").is_err());
+///
+/// assert!(
+///     Title::from_url("https://de.wikipedia.org/wiki/Breil/Brigels").unwrap() !=
+///     Title::from_url("https://de.wikipedia.org/wiki/Breil").unwrap()
+/// );
+/// ```
+#[derive(Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Title {
+    lang: String,
+    name: String,
+}
+
+impl Display for Title {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}:{}", self.lang, self.name)
+    }
+}
+
+impl Title {
+    fn normalize_title(title: &str) -> String {
+        // TODO: Compare with map generator url creation, ensure covers all cases.
+        title.trim().replace(' ', "_")
+    }
+
+    // https://en.wikipedia.org/wiki/Article_Title/More_Title
+    pub fn from_url(url: &str) -> anyhow::Result<Self> {
+        let url = Url::parse(url.trim())?;
+
+        let (subdomain, host) = url
+            .host_str()
+            .ok_or_else(|| anyhow!("Expected host"))?
+            .split_once('.')
+            .ok_or_else(|| anyhow!("Expected subdomain"))?;
+        let host = host.strip_prefix("m.").unwrap_or(host);
+        if host != "wikipedia.org" {
+            bail!("Expected wikipedia.org for domain")
```

Comment:
Does it make sense to print wrong hosts in a log to fix/support them?

Reply:
They are caught at a higher level and logged/saved with the full string